examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1673 @@
+                discriminator_optimizer.zero_grad(set_to_none=True)
+
+                # 1. Decode real and fake (generated) latents back to pixel space.
+                # NOTE: the paper doesn't mention this explicitly AFAIK but I think this makes sense since the
+                # pretrained feature network for the discriminator operates in pixel space rather than latent space.
+                real_image = vae.decode(latents).sample
+                student_gen_image = vae.decode(student_x_0).sample
+
+                # 2. Get discriminator real/fake outputs on the real and fake (generated) images respectively.
+                disc_output_real = discriminator(real_image, prompt_embeds)
+                disc_output_fake = discriminator(student_gen_image.detach(), prompt_embeds)
+
+                # 3. Calculate the discriminator real adversarial loss terms.
+                d_logits_real = disc_output_real.logits
+                # Use hinge loss (see section 3.2, Equation 3 of paper)
+                d_adv_loss_real = torch.mean(F.relu(torch.ones_like(d_logits_real) - d_logits_real))
+
+                # 4. Calculate the discriminator R1 gradient penalty term with respect to the gradients from the real
+                # data.
+                d_r1_regularizer = 0
+                for k, head in discriminator.heads.items():
+                    head_grad_params = torch.autograd.grad(
+                        outputs=d_adv_loss_real, inputs=head.params(), create_graph=True
+                    )
+                    head_grad_norm = 0
+                    for grad in head_grad_params:
+                        head_grad_norm += grad.abs().sum()
+                    d_r1_regularizer += head_grad_norm
+
+                d_loss_real = d_adv_loss_real + args.discriminator_r1_strength * d_r1_regularizer
+                accelerator.backward(d_loss_real, retain_graph=True)
+
+                # 5. Calculate the discriminator fake adversarial loss terms.
+                d_logits_fake = disc_output_fake.logits
+                # Use hinge loss (see section 3.2, Equation 3 of paper)
+                d_adv_loss_fake = torch.mean(F.relu(torch.ones_like(d_logits_fake) + d_logits_fake))
+                accelerator.backward(d_adv_loss_fake)
+
+                d_total_loss = d_loss_real + d_adv_loss_fake
+
+                if accelerator.sync_gradients:
+                    accelerator.clip_grad_norm_(discriminator.parameters(), args.max_grad_norm)
+                discriminator_optimizer.step()
+                discriminator_lr_scheduler.step()
+
+                ############################
+                # 9. Generator Loss
+                ############################
+                optimizer.zero_grad(set_to_none=True)
+
+                # 1. Rerun the disc on generated image, but this time allow gradients to flow through the generator
+                disc_output_fake = discriminator(student_gen_image, prompt_embeds)
```

Comment:
This term is already computed above. Do we need to recompute it here ? Not sure because here we are not doing vanilla GAN training so we might as well be able to utilise that.